nssa/program_methods/guest/src/bin/privacy_preserving_circuit.rs

```diff
@@ -160,28 +163,42 @@ fn main() {
 
                 if visibility_mask[i] == 1 {
                     // Private account with authentication
-                    let (nsk, membership_proof) =
-                        private_auth_iter.next().expect("Missing private auth");
+                    let nsk = private_nsks_iter.next().expect("Missing nsk");
 
                     // Verify the nullifier public key
                     let expected_npk = NullifierPublicKey::from(nsk);
                     if &expected_npk != npk {
                         panic!("Nullifier public key mismatch");
                     }
 
-                    // Compute commitment set digest associated with provided auth path
-                    let commitment_pre = Commitment::new(npk, &pre_states[i].account);
-                    let set_digest = compute_digest_for_path(&commitment_pre, membership_proof);
-
                     // Check pre_state authorization
                     if !pre_states[i].is_authorized {
                         panic!("Pre-state not authorized");
                     }
 
-                    // Compute update nullifier
-                    let nullifier = Nullifier::for_account_update(&commitment_pre, nsk);
+                    let membership_proof_opt = private_membership_proofs_iter
+                        .next()
+                        .expect("Missing membership proof");
+                    let (nullifier, set_digest) = membership_proof_opt
+                        .as_ref()
+                        .map(|membership_proof| {
+                            // Compute commitment set digest associated with provided auth path
+                            let commitment_pre = Commitment::new(npk, &pre_states[i].account);
+                            let set_digest =
+                                compute_digest_for_path(&commitment_pre, membership_proof);
+
+                            // Compute update nullifier
+                            let nullifier = Nullifier::for_account_update(&commitment_pre, nsk);
+                            (nullifier, set_digest)
+                        })
+                        .unwrap_or_else(|| {
+                            // Compute initialization nullifier
+                            let nullifier = Nullifier::for_account_initialization(npk);
+                            (nullifier, DUMMY_COMMITMENT_HASH)
```

Comment:
When initializing private accounts without authentication we are checking that the pre state account is default. I think we need that check here too. 
Otherwise I think I could deploy a dummy program A that does a noop on a single account and execute it privately with authentication with some crafted account owned by another program B (therefore injecting state into that program). This would pass the well beheaved program rules because it does nothing. And so the circuit needs to check that any new initialized private account comes from an uninitialized pre state.

Maybe a good test to add!

Reply:
Added, thanks. Took me some time to understand how to trigger this. Actually even without this check malicious transaction wouldn't pass on-chain checks, as it will have the same nullifier which has been already seen.

Reply:
Ah nice thanks